ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarRecoveringCustodyImpl.java

```diff
@@ -298,29 +275,37 @@ public SafeFuture<Optional<DataColumnSidecar>> getCustodyDataColumnSidecarByRoot
 
   @Override
   public SafeFuture<Void> onNewValidatedDataColumnSidecar(
-      final DataColumnSidecar dataColumnSidecar) {
-    createOrUpdateRecoveryTaskForDataColumnSidecar(
-        DataColumnSlotAndIdentifier.fromDataColumn(dataColumnSidecar));
-    return delegate.onNewValidatedDataColumnSidecar(dataColumnSidecar);
+      final DataColumnSidecar dataColumnSidecar, final RemoteOrigin remoteOrigin) {
+    // Recovery is not needed for locally produced or recovered data,
+    // we will get everything for it in custody w/o reconstruction
+    if (remoteOrigin.equals(RemoteOrigin.RPC) || remoteOrigin.equals(RemoteOrigin.GOSSIP)) {
+      createOrUpdateRecoveryTaskForDataColumnSidecar(dataColumnSidecar);
+    }
+    return delegate.onNewValidatedDataColumnSidecar(dataColumnSidecar, remoteOrigin);
   }
 
-  private synchronized void createOrUpdateRecoveryTaskForDataColumnSidecar(
-      final DataColumnSlotAndIdentifier identifier) {
-    if (recoveryTasks.containsKey(identifier.getSlotAndBlockRoot())) {
-      final RecoveryTask existing = recoveryTasks.get(identifier.getSlotAndBlockRoot());
-      existing.existingColumnIds().add(identifier);
-      maybeStartRecovery(existing);
-    } else {
-      final ConcurrentHashSet<DataColumnSlotAndIdentifier> identifiers = new ConcurrentHashSet<>();
-      identifiers.add(identifier);
+  private void createOrUpdateRecoveryTaskForDataColumnSidecar(final DataColumnSidecar sidecar) {
+    if (!recoveryTasks.containsKey(sidecar.getSlotAndBlockRoot())) {
+      final Map<DataColumnSlotAndIdentifier, DataColumnSidecar> sidecars =
+          new ConcurrentHashMap<>();
+      sidecars.put(DataColumnSlotAndIdentifier.fromDataColumn(sidecar), sidecar);
       final RecoveryTask recoveryTask =
           new RecoveryTask(
-              new AtomicReference<>(null),
-              identifiers,
+              sidecar.getSlotAndBlockRoot(),
+              sidecars,
               new AtomicBoolean(false),
               new AtomicBoolean(false));
-      recoveryTasks.put(identifier.getSlotAndBlockRoot(), recoveryTask);
```

Comment:
i still think this is much better expressed as putIfAbsent, im not sure why we're so set on containsKey.
if its only because of the sidecars concurrentHashMap just abstract that away anddo the `new RecoveryTask` inline in the putIfAbsent